File: Assignments/A09/2.c
#include <stdio.h>
#include <stdlib.h>

int find(int parent[], int i) {
    if (parent[i] != i) {
        parent[i] = find(parent, parent[i]);
    }
    return parent[i];
}

void unionSet(int parent[], int rank[], int i, int j) {
    int iRoot = find(parent, i);
    int jRoot = find(parent, j);
    if (iRoot == jRoot) {
        return;
    }
    if (rank[iRoot] < rank[jRoot]) {
        parent[iRoot] = jRoot;
    } else if (rank[iRoot] > rank[jRoot]) {
        parent[jRoot] = iRoot;
    } else {
        parent[jRoot] = iRoot;
        rank[iRoot]++;
    }
}

int main() {
    int n, q;
    scanf("%d %d", &n, &q);

    int *parent = malloc((n+1) * sizeof(int));
    int *rank = calloc(n+1, sizeof(int));

    for (int i = 0; i <= n; i++) {
        parent[i] = i;
    }

    for (int i = 0; i < q; i++) {
        int op_type, x, y;
        scanf("%d %d %d", &op_type, &x, &y);

        if (op_type == 1) {
            unionSet(parent, rank, x, y);
        } else if (op_type == 2) {
            int root_x = find(parent, x);
            for (int j = x + 1; j <= y; j++) {
                unionSet(parent, rank, root_x, j);
            }
        } else if (op_type == 3) {
            if (find(parent, x) == find(parent, y)) {
                printf("YES\n");
            } else {
                printf("NO\n");
            }
        }
    }

    free(parent);
    free(rank);

    return 0;
}

File: Assignments/A09/2.py
def find(parent, i):
    if parent[i] != i:
        parent[i] = find(parent, parent[i]) 
    return parent[i]

def union(parent, rank, i, j):
    i_root = find(parent, i)
    j_root = find(parent, j)
    if i_root == j_root:
        return
    if rank[i_root] < rank[j_root]:
        parent[i_root] = j_root
    elif rank[i_root] > rank[j_root]:
        parent[j_root] = i_root
    else:
        parent[j_root] = i_root
        rank[i_root] += 1

n, q = map(int, input().split())
parent = [i for i in range(n+1)]
rank = [0] * (n+1)

for _ in range(q):
    op_type, x, y = list(map(int, input().split()))
    if op_type == 1 and x!=y:
        union(parent, rank, x, y)
    elif op_type == 2:
        for i in range(x, y):
            union(parent, rank, i, i+1)
    elif op_type == 3:
        if find(parent, x) == find(parent, y):
            print("YES")
        else:
            print("NO")